astropy/wcs/wcs.py

```diff
@@ -2959,8 +2959,18 @@ def slice(self, view, numpy_order=True):
                 else:
                     wcs_new.wcs.crpix[wcs_index] -= iview.start
 
-            nitems = len(builtins.range(self._naxis[wcs_index])[iview])
-            wcs_new._naxis[wcs_index] = nitems
+            try:
+                # range requires integers but the other attributes can also
+                # handle arbitary values, so this needs to be in a try/except.
+                nitems = len(builtins.range(self._naxis[wcs_index])[iview])
+            except TypeError as exc:
+                if 'indices must be integers' not in str(exc):
+                    raise
+                warnings.warn("NAXIS{0} could not be updated because one or "
+                              "multiple indices ('{1}') were not integral."
+                              "".format(wcs_index, iview), AstropyUserWarning)
```

Comment:
OK, I understand your intent now.  However, this will return the original WCS unedited, which I believe is not the now-deprecated behavior; I believe it would accept floats before and just truncated them (i.e., take `int(slice)`).  But I'm not certain of that...